## A7: Final Project Report

### I. Introduction

For years researchers have explored and tried to map the way we can tap into human emotional responses with simple environmental cues, such as the visual surroundings, auditory stimuli, sensory stimuli, etc. 

There has been significant research in the field of identifying sounds as purely belonging to a certain sound group (i.e. happy or sad) which has resulted in the creation of a corpus of melodies, created specifically as a building block to delve deeper into how human beings map emotions and music.

Such studies help us further understand how our brains connect memories, emotions, and external stimuli together.

This project tries to explore certain questions in the area of mapping musical stimuli with first emotional response, using data generated by a study conducted in Stockholm and Finland in 2013 titled: [Emotional expression in music: contribution, linearity, and additivity of primary musical cues](https://www.frontiersin.org/articles/10.3389/fpsyg.2013.00487/full)

The data is licensed under [CC0](https://creativecommons.org/publicdomain/zero/1.0/) (public use) and is hosted by [Harvard Dataverse](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/IFOBRN)

## II. Related work and Background

The corpus of work that inspired this project is

### a) [Happy, sad, scary and peaceful musical excerpts for research on emotions - Veillard et. al](https://www.researchgate.net/publication/247496910_Happy_sad_scary_and_peaceful_musical_excerpts_for_research_on_emotions)

This body of work categorized 56 musical excerpts as Happy, Sad, Scary, and Peaceful on a continuous range (assigning a percentage wise score for each melody in each category)

This work generated 4 key melodies that were classified as unambigously Sad, Scary, Happy, or Peaceful, and these melodies were used as the basis for the research that generated the data I have used in this project.

This research was conducted with the key goal in mind of setting a stepping stone for further research in the area of emotion and melody mapping.

This study cites several other papers that map emotion conveyed by a melody to be non-subjective, but rather very clearly common in musically educated and uneducated listeners ( [Bigand et al](http://leadserv.u-bourgogne.fr/files/publications/000103-multidimensional-scaling-of-emotional-responses-to-music-the-effect-of-musical-expertise-and-of-the-duration-of-the-excerpts.pdf) )

The purpose of this study was not directed in the area of mapping emotional response to CHANGE in melody, but to see if listeners could map unlabelled music intended to convey a certain emotion as the same, and the found inherently that that was in fact the case. Happy and Sad music was easily identified as such by an overwhelming majority.


### b) [Emotion rendering in music: range and characteristic values of seven musical variables](https://www.ncbi.nlm.nih.gov/pubmed/21696717)

This body of work identifies the characteristics of music that actually affect the emotional response of a listener. As opposed to factorially altering all characteristics of music with arbitrary values, this research proposes that only certain values of certain characteristics actyally illicit an emotional response.

This work is on the basis of which the data on the 200 melodic stimuli I have used in this project was designed. 

### c) [Emotional expression in music: contribution, linearity, and additivity of primary musical cues](https://www.frontiersin.org/articles/10.3389/fpsyg.2013.00487/full)

This body of work generated the data that I have used in this project. The researchers used previously identified unabiguously sad happy peaceful and scary melodies (Using Veillard et al. results) to generate 200 musical stimuli, by factorially changing key values as identified by Bresin et. al. The researchers then asked 46 test subjects to listen to this stimuli and rate each melody as Happy, Scary, Sad, and Peaceful on a scale of 1 to 7 (Thus each melody would have 4 ratings per user.)

From this data the researchers identified the linearity of the relationship between emotional response and each characteristic, and tried to break down the dependency of each value and the emotional response.

## III. Research Questions, Results, and Code

The questions I wanted to focus on are a small subset of what Tuomas et al. tried to look into, I wanted to conduct some Exploratory Data Analysis to see how the relationship between certain characteristics of music aligned with my own understanding of music.

### Setup

First let us import the relevant packages and load the data:

    

In [3]:
import numpy as np
import pandas as pd

from sklearn.linear_model import LinearRegression

In [4]:
design_mat = pd.read_csv('../Data/design_matrix.csv',index_col=0)

In [5]:
emotion_rating = pd.read_csv('../Data/mean_emotion_ratings.csv',index_col=0)

### Data

Before we delve into the research questions, let's take a quick look at the data

#### Design Matrix

This table contains information on 200 melodic stimuli, with each row describing one melody.

The columns contain the following encodings:

- **Register** - 6 levels
    - 1: 53 MIDI pitch
    - 2: 59 MIDI pitch
    - 3: 65 MIDI pitch
    - 4: 71 MIDI pitch
    - 5: 77 MIDI pitch
    - 6: 83 MIDI pitch
- **Mode** - 2 modes
    - 1: Major key
    - 2: Minor key
- **Tempo** - 5 tempos (in NPS)
    - 1: 1.2
    - 2: 2
    - 3: 2.8
    - 4: 4.4
    - 5: 6
- **Sound level** - 5 sound levels (in dB)
    - 1: -10
    - 2: -5
    - 3: 0
    - 4: +5
    - 5: +10
- **Articulation** - 4 levels (from legato to staccato)
- **Timbre** - 3 levels 
    - 1: trumpet
    - 2: flute
    - 3: horn
- **Melody** - 4 types
    - 1: Sad
    - 2: Scary
    - 3: Happy
    - 4: Peaceful
    

Looking at the first five rows of the design_mat data we see how these encodings would look:

In [6]:
design_mat.head()

Register  Mode  Tempo  Soundlevel  Articulation  Timbre  Melody
Nro                                                                 
1           4     1      4           4             2       2       4
2           5     1      4           1             1       2       2
3           2     2      5           1             1       2       1
4           1     1      5           4             4       1       2
5           3     2      1           3             2       2       1

#### Mean emotional response

This table contains the mean emotional response for 200 melodic stimuli, with each row co-inciding with one melody. This mean response has been calculated by averaging out the reaction of 46 users for each melody and each emotional category.

Looking at the first five rows of the mean emotional response table, we can see what that would look like:

In [7]:
emotion_rating.head()

Scary   Happy     Sad  Peaceful
Nro                                  
1    1.2889  4.4667  1.7111    3.1333
2    1.0667  5.4444  1.4889    4.4889
3    2.0222  1.4889  3.7778    2.7111
4    2.2889  4.1111  1.2667    1.4889
5    1.4000  1.4667  5.0444    3.8444

## Research Question 1: How linear is the relationship between "Melody" and "Mean Emotional Response"?

    Since the melody is already classified in the same buckets as Mean Emotional Response, it would be interesting to know if the overreaching melody type has a predominant affect on the mean emotional response

To analyze this, first for each melody I assigned a "projected emotion" value in the mean emotional response table. This value represented the max mean emotional response for each melody, and assigned the same encoding as that of "Melody" from teh design_mat table to the stimuli, 1: Sad, 2: Scary, 3: Happy, 4: Peaceful

In [8]:
a = []

for i in emotion_rating.values:
    if max(i) == i[0]:
        a.append(2)
    elif max(i) == i[1]:
        a.append(3)
    elif max(i) == i[2]:
        a.append(1)
    elif max(i) == i[3]:
        a.append(4)
    

In [9]:
emotion_rating["projected emotion"] = a

In [10]:
emotion_rating.head()

Scary   Happy     Sad  Peaceful  projected emotion
Nro                                                     
1    1.2889  4.4667  1.7111    3.1333                  3
2    1.0667  5.4444  1.4889    4.4889                  3
3    2.0222  1.4889  3.7778    2.7111                  1
4    2.2889  4.1111  1.2667    1.4889                  3
5    1.4000  1.4667  5.0444    3.8444                  1

Thus we see that now each melody has a projected emotion. Now I calculated how many times this projected emotion matched the "Melody" encoding in the design_mat table

In [11]:
count = 0
for i in range(len(design_mat)):
    if design_mat.iloc[i,6] == emotion_rating.iloc[i,4]:
               count = count + 1

In [12]:
count/len(design_mat)

0.23


### Thus 23% stimuli match projected emotion and actual theme of the melody

This however still does not answer our question. To answer that we will plot the percentage wise matching of each stimuli's "Melody" and "projected emotion"

First we will store a copy of the calculated "projected emotion" in the design_mat table as "overall_emotion"

In [13]:
design_mat["overall_emotion"] = emotion_rating["projected emotion"]

We will now calculate each emotion's match percentage

In [15]:
#Matching count for each Melody and overall_emotion
mel1 = 0 
mel2 = 0
mel3 = 0
mel4 = 0

#Total count for each Melody
tot1 = 0
tot2 = 0
tot3 = 0
tot4 = 0

for i in design_mat.values:
    if i[6] == i[7]:
        if i[6] == 1:
            mel1 = mel1 + 1
        elif i[6] == 2:
            mel2 = mel2 + 1
        elif i[6] == 3:
            mel3 = mel3 + 1
        elif i[6] == 4:
            mel4 = mel4 + 1
    
    if i[6] == 1:
        tot1 = tot1 +1
    elif i[6] == 2:
        tot2 = tot2 + 1
    elif i[6] == 3:
        tot3 = tot3 +1
    else:
        tot4 = tot4 + 1
            
           
    

In [28]:
mel1p = mel1*100/tot1

In [29]:
mel2p = mel2*100/tot2

In [30]:
mel3p = mel3*100/tot3

In [31]:
mel4p = mel4*100/tot4

We will store these values in a dataframe for ease of plotting

In [32]:
df = pd.DataFrame(data=[["Sad",mel1p],
                       ["Scary",mel2p],
                       ["Happy", mel3p],
                       ["Peaceful",mel4p]], columns= ["Melody Emotion","Matching Percentage"])

In [ ]:
import plotly.express as px

fig = px.bar(df, x='Melody Emotion', y='Matching Percentage')
fig.update_traces(marker_color = "purple")
fig.show()

## Resulting graph:

<img src="RQ1.png">

### Thus we can see that Melody and overall_response do not match very much, thus the characteristics we have factorially altered does have some impact on mean emotional response

## Research Question 2: Which factor most affects the relationship between "Melody" and "Mean Emotional Response"?

    Which of the six musical factors most changes the mean emotional response from co-inciding with the Melody

This should help us identify which factor most alters someone view on a melody. To calculate this we shall perform some simple linear regression and look at the intercept values to see which characteristic paired with melody reduces its impact on overall response

Let us first see what the coefficient is for Linear regression of Mean emotional Response on Meldoy without any confounding variables


In [34]:
X = np.array(design_mat.Melody).reshape((-1, 1))

X.shape

y = np.array(design_mat.overall_emotion)

y.shape

model = LinearRegression()
model.fit(X,y)
print("coeff: ",model.coef_)

coeff:  [-0.01130298]


The coefficient is -0.0113, now any characteristic that changes this value, making it closer to zero (Reducing its importance on the mean emotional response is the characteristic that most affects the relationship between melody and mean emotional response.

In [35]:
X = design_mat[['Register', 'Melody']].values

model = LinearRegression()
model.fit(X,y)


df = pd.DataFrame(data=X,columns=["Register","Melody"])

coeff_df = pd.DataFrame(model.coef_, df.columns, columns=['Coefficient'])  
coeff_df

Coefficient
Register     0.131411
Melody      -0.012788

In [36]:
X = design_mat[['Timbre', 'Melody']].values

model = LinearRegression()
model.fit(X,y)


df = pd.DataFrame(data=X,columns=["Timbre","Melody"])

coeff_df = pd.DataFrame(model.coef_, df.columns, columns=['Coefficient'])  
coeff_df

Coefficient
Timbre     0.007490
Melody    -0.011272

In [37]:
X = design_mat[['Soundlevel', 'Melody']].values

model = LinearRegression()
model.fit(X,y)


df = pd.DataFrame(data=X,columns=["Soundlevel","Melody"])

coeff_df = pd.DataFrame(model.coef_, df.columns, columns=['Coefficient'])  
coeff_df

Coefficient
Soundlevel    -0.061485
Melody        -0.011290

In [38]:
X = design_mat[['Articulation', 'Melody']].values

model = LinearRegression()
model.fit(X,y)


df = pd.DataFrame(data=X,columns=["Articulation","Melody"])

coeff_df = pd.DataFrame(model.coef_, df.columns, columns=['Coefficient'])  
coeff_df

Coefficient
Articulation     0.019712
Melody          -0.012095

In [39]:
X = design_mat[['Mode', 'Melody']].values

model = LinearRegression()
model.fit(X,y)


df = pd.DataFrame(data=X,columns=["Mode","Melody"])

coeff_df = pd.DataFrame(model.coef_, df.columns, columns=['Coefficient'])  
coeff_df

Coefficient
Mode      -1.431565
Melody    -0.011603

In [40]:
X = design_mat[['Tempo', 'Melody']].values

model = LinearRegression()
model.fit(X,y)


df = pd.DataFrame(data=X,columns=["Tempo","Melody"])

coeff_df = pd.DataFrame(model.coef_, df.columns, columns=['Coefficient'])  
coeff_df

Coefficient
Tempo      0.056447
Melody    -0.009194

### Looking at the above values, we can see that Tempo has the most effect on the relationship between Melody and Overall emotion

## Research Question 3: Does a "minor" mode mostly illicit a "Scary" emotional response?

    As a listener of music generally songs in a minor key can come off as creepy, I would like to know if that holds true for the populous


For this analysis we will first find the subset of design_mat table that only contains Mode: 2 (Minor Key)
    


In [41]:
minor_key = design_mat[design_mat["Mode"]==2]

In [42]:
minor_key.head()

Register  Mode  Tempo  Soundlevel  Articulation  Timbre  Melody  \
Nro                                                                    
3           2     2      5           1             1       2       1   
5           3     2      1           3             2       2       1   
6           2     2      3           2             4       2       2   
7           1     2      2           1             3       1       4   
9           1     2      1           4             4       1       1   

     overall_emotion  
Nro                   
3                  1  
5                  1  
6                  4  
7                  1  
9                  1

Looking at the above table we see that only Mode 2 melodies are in this table, we shall remove the mode column now as it has no relevance for the rest of this analysis

In [43]:
del minor_key["Mode"]

We will now calculate the percentage of each overall_emotion in the minor_key table

In [46]:
count_sad = 0
count_happy = 0
count_peaceful = 0
count_scary = 0

for i in minor_key.values:
    if i[6] == 1:
        count_sad = count_sad + 1
    elif i[6] == 2:
        count_scary = count_scary + 1
    elif i[6] == 3:
        count_happy = count_happy + 1
    elif i[6] == 4:
        count_peaceful = count_peaceful + 1
    
        

In [47]:
df = pd.DataFrame(data = [["Sad", (count_sad*100)/len(minor_key)],
                          ["Happy",(count_happy*100)/len(minor_key)],
                          ["Peaceful",(count_peaceful*100)/len(minor_key)],
                          ["Scary",(count_scary*100)/len(minor_key)]],
                 columns = ["Emotion", "%ge of minor key first response"] )


In [ ]:
import plotly.express as px

fig = px.bar(df, x='Emotion', y='%ge of minor key first response')
fig.show()

## Resulting graph:

<img src="RQ2.png">

### This shows that most minor key songs are categorized as sad, as opposed to scary. To dig deeper into this let us see how mean emotional response for scary maps to the "Mode" column of the design_mat (Which stores the key information for each stimuli)

We will add a column to store the mean emotional response under "Scary" from the emotional_rating table to the design_mat table

In [49]:
design_mat["average scary"] = emotion_rating["Scary"]

In [ ]:
import plotly.express as px

fig = px.scatter(design_mat, x="average scary", y="Mode",
	         color = "overall_emotion", color_continuous_scale = "RdBu")
fig.update_traces(mode='markers',  marker_size=10)
fig.show()

## Resulting graph:

<img src="RQ3.png">

### Thus we see that as the Mode is set to 2 (Minor Key), the points are mostly warm colors (implying Sad or Scary) even if they are scored low on the average scary scale. 

## Research Question 4 : Does a high tempo lead to a happier emotional response?

    Generally speaking fast tempo-ed songs (such as pop music) generate a positive response, it would be interesting to know if that is true here as well

To analyze this we will follow a similar process, mapping Tempo against average_happy rating.

Let us first get the average happy value into the design_mat table from emotional_rating table

In [60]:
design_mat["average happy"] = emotion_rating["Happy"]

To add another dimension to our analysis, we shall count how many melodies have matching tempos and matching average rating (just to see if any one tempo has an overwhelmingly common average happy response)

In [62]:
design_mat["count"] = 1 # Assign a count to each melody

In [63]:
# Now group by Tempo and average happy summing the count of each occurence 
anotherdf = pd.DataFrame(data = design_mat.groupby(["Tempo","average happy"],as_index=False)["count"].sum(), 
                         columns= ["Tempo","average happy","count"])

In [68]:
anotherdf.head()

Tempo  average happy  count
0      1         1.1111      2
1      1         1.1333      1
2      1         1.1556      2
3      1         1.1778      2
4      1         1.2000      1

Thus this table keeps track of each tempo, averahe happy score pair and their individual occurence count, plotting this we see that:

In [ ]:
import plotly.express as px

fig = px.scatter(anotherdf, x="average happy", y="Tempo",
	         size="count")
fig.show()

## Resulting graph:

<img src="RQ4.png">

### The lower the tempo is, the lower the average happy score is. Higher the tempo is, the more spread out the score is across the board. Thus this would imply that a low tempo does imply a non-happy response, but a high tempo does not inherently imply a happy emotional response

## IV. Discussion - Limitations and Future Work

- Although this data does provide proof in the form of previous research, that a complete factorial stimuli generation is unnecessary, it would be more interesting to have a complete and rich dataset to properly analyze any relationships that may have been missed or could lead to a deeper understanding of how music and emotions are mapped

- Mean emotional response is a good descriptor to record, but providing us with each response of the 46 individuals could have let us analyze more things that could have conncted each user / other descriptors could have beem amalyzed such as median or mode to see if these provided different insights

- Again although the research is backed by previous research that emotional response is not subjective, it could have been interesting to spread the test users across a more diverse geographic location and included some more demographic information such as age or occupation, to see if any more relationshups could have been mapped.

- Also the test situation for each user has not been very clearly outlined in the study (I should imagine if I'm listening to 200 musical stimuli at a stretch, all musical stimuli would begin to sound sad)

- For future work it would be interesting to see more in depth analysis of both testers and the stimuli to gain further insight on how emotions are related to melodic stimuli

## V. Conclusions and Reflections

It is clear that there are some definite relationships between musical stimuli and emotional response, and further more it is clear that certain characteristics of musical stimuli can be altered to generate or incur a specific emotional response, such an insight can lead to more in deoth understadning of how emotions are mapped in our brains, but could also potentially be a cause for concern if used in unethical situations in illicit a specific response. It feels like such research could steer dabgerously close to topics such a brain washing or hypnosis, which has it's own set of ethical dilemmas and conundrums. 

## VI. References

- [Happy, sad, scary and peaceful musical excerpts for research on emotions - Veillard et. al](https://www.researchgate.net/publication/247496910_Happy_sad_scary_and_peaceful_musical_excerpts_for_research_on_emotions)
- [Emotion rendering in music: range and characteristic values of seven musical variables - Bresin et. al](https://www.ncbi.nlm.nih.gov/pubmed/21696717)
-  [Emotional expression in music: contribution, linearity, and additivity of primary musical cues - Eerola et al.](https://www.frontiersin.org/articles/10.3389/fpsyg.2013.00487/full)
- [Multidimensional scaling of emotional responses to
music: The effect of musical expertise and of the
duration of the excerpts - Bigand et al](http://leadserv.u-bourgogne.fr/files/publications/000103-multidimensional-scaling-of-emotional-responses-to-music-the-effect-of-musical-expertise-and-of-the-duration-of-the-excerpts.pdf)

(Please Note: References to each cited research has been peppered through out the assignment)